# Step 1 — Import Required Libraries and Load the Dataset

We begin by importing the necessary Python libraries and loading the raw social media dataset for inspection and cleaning.

In [138]:
import pandas as pd
import numpy as np
import re
#embedding:
from sentence_transformers import SentenceTransformer
#vector data base setup:
from qdrant_client import QdrantClient
from qdrant_client.http.models import VectorParams, Distance, PointStruct, models

#cosine similarity:
from sklearn.metrics.pairwise import cosine_similarity

#gemini:
import google.generativeai as genai

#agents:
from agno.agent import Agent
from agno.models.google import Gemini
from agno.tools.thinking import ThinkingTools
from agno.tools.reasoning import ReasoningTools

#sandbox : 
import modal

#front-end library:
import gradio as gr

In [139]:
df = pd.read_csv("social_media_analyzed.csv")
df["post_text"] = df.groupby("post_id")["text"].transform("first")
df

post_id         created_time  \
0      7.429425e+18  2024-11-06 18:59:24   
1      7.438736e+18  2024-11-18 23:09:03   
2      7.438736e+18  2024-11-19 13:26:55   
3      7.440197e+18  2024-11-22 21:37:41   
4      7.440197e+18  2024-11-22 21:41:03   
...             ...                  ...   
20507  2.175670e+15  2025-04-08 19:14:28   
20508  2.175670e+15  2025-04-08 20:10:41   
20509  2.175670e+15  2025-04-10 16:59:06   
20510  3.800919e+15  2024-04-14 18:18:49   
20511  2.175670e+15  2025-04-08 19:10:04   

                                                    text  \
0                                      سريتو ب 2dh واعرر   
1       ibtissam akhti kinder wach masmo7a 3afak jawbini   
2                          Crunch hta hiya fiha el3alama   
3                                              ميريندينا   
4                                               مريندينا   
...                                                  ...   
20507                                                NaN   
20508                                                NaN   
20509                                                NaN   
20510  آااااه يا ليام لعزيز اعليك تحط ليه الصينية و ا...   
20511                                                NaN   

                      type_of_snack activity_context temporal_context  \
0                       [['snack']]               []               []   
1                       [['snack']]               []               []   
2                       [['snack']]               []               []   
3                       [['snack']]               []               []   
4                       [['snack']]               []               []   
...                             ...              ...              ...   
20507                   [['snack']]               []               []   
20508                   [['snack']]               []               []   
20509  [['snack', 'Healthy snack']]      [['OTHER']]      [['OTHER']]   
20510                   [['snack']]     [['UNSURE']]     [['UNSURE']]   
20511                   [['snack']]               []               []   

              spatial_context                          social_context_detail  \
0                          []                                             []   
1                          []                                             []   
2                          []                                             []   
3                          []                                             []   
4                          []                                             []   
...                       ...                                            ...   
20507                      []                                             []   
20508                      []                                             []   
20509  [['At Home', 'OTHER']]  [['With Children / Family', 'Solo', 'OTHER']]   
20510            [['UNSURE']]                                   [['UNSURE']]   
20511                      []                                             []   

                                functional_triggers  \
0                                                []   
1                                                []   
2                                                []   
3                                                []   
4                                                []   
...                                             ...   
20507                                            []   
20508                                            []   
20509  [['Hunger gap / tide over', 'Ritual habit']]   
20510                            [['Ritual habit']]   
20511                                            []   

                                   emotional_triggers  \
0                                                  []   
1                                                  []   
2                                                  []   
3                                  

In [199]:
df.head()

post_id         created_time  \
0  7.429425e+18  2024-11-06 18:59:24   
1  7.438736e+18  2024-11-18 23:09:03   
2  7.438736e+18  2024-11-19 13:26:55   
3  7.440197e+18  2024-11-22 21:37:41   
4  7.440197e+18  2024-11-22 21:41:03   

                                             text type_of_snack  \
0                                سريتو ب dh واعرر   [['snack']]   
1  ibtissam akhti kinder wach masmoa afak jawbini   [['snack']]   
2                    crunch hta hiya fiha elalama   [['snack']]   
3                                       ميريندينا   [['snack']]   
4                                        مريندينا   [['snack']]   

  activity_context temporal_context spatial_context social_context_detail  \
0               []               []              []                    []   
1               []               []              []                    []   
2               []               []              []                    []   
3               []               []              []                    []   
4               []               []              []                    []   

  functional_triggers emotional_triggers consumer_profiles  \
0                  []                 []                []   
1                  []                 []                []   
2                  []                 []                []   
3                  []                 []                []   
4                  []                 []                []   

                      barriers nutrition_profile  preparation_effort  yyyy_mm  \
0  [['Price / Affordability']]                []  [['Ready-to-Eat']]  2024-11   
1                 [['UNSURE']]                []  [['Ready-to-Eat']]  2024-11   
2                           []                []  [['Ready-to-Eat']]  2024-11   
3                           []                []  [['Ready-to-Eat']]  2024-11   
4                           []                []  [['Ready-to-Eat']]  2024-11   

                                          post_text  \
0                                 سريتو ب 2dh واعرر   
1  ibtissam akhti kinder wach masmo7a 3afak jawbini   
2  ibtissam akhti kinder wach masmo7a 3afak jawbini   
3                                         ميريندينا   
4                                         ميريندينا   

                                           full_info  
0  \npost_id: 7.429424947254152e+18\ncomment: سري...  
1  \npost_id: 7.438736352994168e+18\ncomment: ibt...  
2  \npost_id: 7.438736352994168e+18\ncomment: cru...  
3  \npost_id: 7.44019699323167e+18\ncomment: ميري...  
4  \npost_id: 7.44019699323167e+18\ncomment: مرين...

In [140]:
#drop missing values:
df = df.dropna(subset = "text")

#clean text:
df["text"] = df["text"].str.lower()
def cleanText(text):
    text = re.sub(r"http\S+", "", text)  # remove URLs
    text = re.sub(r"@\w+", "", text)     # remove @mentions
    text = re.sub(r"#\w+", "", text)     # remove hashtags
    text = re.sub(r"[^\w\s]", "", text)  # remove punctuation
    text = re.sub(r"\d+", "", text)      # remove numbers
    text = re.sub(r"\s+", " ", text)     # remove extra whitespace
    return text.strip()

df["text"] = df["text"].apply(cleanText)

# Drop any rows where cleaned text is empty
df = df[df['text'].str.strip() != ""]

C:\Users\youss\AppData\Local\Temp\ipykernel_21076\668454729.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["text"] = df["text"].str.lower()
C:\Users\youss\AppData\Local\Temp\ipykernel_21076\668454729.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["text"] = df["text"].apply(cleanText)


In [141]:
df["full_info"] = df.apply(lambda row: f"""
post_id: {row["post_id"]}
comment: {row['text']}
type_of_snack: {row['type_of_snack']}
activity_context: {row['activity_context']}
temporal_context: {row['temporal_context']}
spatial_context: {row['spatial_context']}
social_context: {row['social_context_detail']}
functional_triggers: {row['functional_triggers']}
emotional_triggers: {row['emotional_triggers']}
consumer_profiles: {row['consumer_profiles']}
barriers: {row['barriers']}
nutrition_profile: {row['nutrition_profile']}
preparation_effort: {row['preparation_effort']}
""", axis=1)

In [142]:
df.to_csv("social_media_analyzed.csv")

# Step 2 — Embedding Social Media Posts

The goal of this step is to convert each social media post (along with its context) into a fixed-size vector representation using a pre-trained multilingual model. These embeddings capture the **semantic meaning** of the text and will be used later for **similarity search** via a vector database.

---

## 2.1 Model: distiluse-base-multilingual-cased-v2

We use the `SentenceTransformer` model `distiluse-base-multilingual-cased-v2` from the `sentence-transformers` library. This model is designed to work across multiple languages, including Darija, French, and Arabic.

---


In [ ]:
model = SentenceTransformer("distiluse-base-multilingual-cased-v2")

In [8]:
texts = df["text"].tolist()

In [ ]:
embeddings = model.encode(texts, show_progress_bar= True)

In [ ]:
embeddings.shape

In [ ]:
df["embeddings"]= embeddings.tolist()

## Step 4: Uploading Embeddings to Qdrant

Now that we have generated high-quality sentence embeddings for all social media posts, the next step is to store these vectors in a vector database for efficient semantic search.

We will use **Qdrant Cloud**, a powerful open-source vector database designed for high-speed similarity search and filtering.

### 🎯 Objective

Store each comment's embedding, along with its associated metadata (like  type, barriers, emotional trigger...), in Qdrant so we can later perform fast and meaningful retrievals using cosine similarity.

In [ ]:
client = QdrantClient(
    url = "https://1462be1d-ff96-47a8-87a9-a61c66fd3c84.eu-west-2-0.aws.cloud.qdrant.io",
    api_key= "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3MiOiJtIn0.8dXaklGZ6uMW8nwrIF9cwnOsJp6Z_1R0wcXIWFcbtX8")

In [ ]:
points = []

for i, row in df.iterrows():
    point = PointStruct(
        id=i,
        vector=row["embeddings"],
        payload={
            col: row[col] for col in df.columns if col != "embeddings"
        }
    )
    points.append(point)

In [ ]:
client.recreate_collection(
    collection_name="social_media_feedback",
    vectors_config=models.VectorParams(
        size=512,
        distance=models.Distance.COSINE
    )
)


from tqdm import tqdm

batch_size = 500  # safe value, increase gradually if needed

for i in tqdm(range(0, len(points), batch_size)):
    batch = points[i:i + batch_size]
    client.upsert(
        collection_name="social_media_feedback",
        points=batch
    )

In [ ]:
query = "what are people saying about merendina?"

query_vector = model.encode(query).tolist()

results = client.search(
    collection_name="social_media_feedback",
    query_vector=query_vector,
    limit=5
)

for res in results:
    print(res.payload["text"], res.score)

#  Part 5: Gemini Integration (LLM-Powered Analysis)

In this step, we connect our semantic search engine to **Gemini Pro**, Google's state-of-the-art large language model.

We use Gemini to generate **natural language answers** based on the most relevant social media posts retrieved from Qdrant. This creates a powerful Retrieval-Augmented Generation (RAG) pipeline that can:


In [ ]:
genai.configure(api_key="AIzaSyAYFKWG_34OznV9eMowZ-7kWEx1KBdlurc")

# Load the model (gemini-pro is the text model)
model1 = genai.GenerativeModel("gemini-2.5-flash")


In [ ]:
retrieved_texts = [res.payload["text"] for res in results]
context = "\n".join(retrieved_texts)

prompt = f"""
You're a helpful assistant. Based only on the following social media posts, answer this question:

Question: What do people say about merendina?

Posts:
{context}

Answer:
"""

response = model1.generate_content(prompt)
print(response.text)

In [ ]:
def answer_question(question, top_k = 10):
    try:
        q_vector = model.encode(question).tolist()
        results = client.search(collection_name="social_media_feedback",query_vector=q_vector,limit=top_k)
        if not results:
            return "no relevant post found"
        enriched_examples = []

        for res in results:
            # Get the comment and its post_id
            target_comment = res.payload["text"]
            post_id = res.payload["post_id"]

            # Get all comments for this post_id
            group = df[df["post_id"] == post_id]
            if group.empty:
                continue  # Skip if no comments found for this post_id

            # Get the post text
            post_text = group["post_text"].iloc[0]

            # Get other comments under the same post (excluding the target comment)
            other_comments = group[group["text"] != target_comment]["text"].tolist()
            other_context = "\n".join(other_comments)

            # Build prompt for this comment
            example = f"""
Post:
{post_text}

Other Comments:
{other_context}

Target Comment:
{target_comment}
"""
            enriched_examples.append(example)

        # Merge all contextual examples into one block
        full_context = "\n---\n".join(enriched_examples)

        prompt = f""" You are a senior consumer-insights analyst.

        You will receive:
        Question:  a natural-language query from the user.  
        Comment : a short list of social-media comments already retrieved for you (and the post it belongs to for context). 
        They may be in Arabic, Darija, French, or English.  
        They may contain slang, spelling mistakes, emojis, or code-switching.
        A specific user comment from each thread.
        Your task is to analyze each user's comment in the context of the post and the other comments.
        Your task
        1. Read all comments carefully. Use only these comments and their post for evidence and context.  
        2. Answer the user's question as clearly and concisely as possible.  
        3. For each target comment, also identify the main emotion expressed.Don't mention them, use them for extra analysis. Choose from:
            - Joy / Satisfaction
            - Anger / Frustration
            - Sadness / Disappointment
            - Surprise / Shock
            - Fear / Anxiety
            - Confusion
            - Love / Attachment
            - Disgust
            - Humor / Sarcasm
            - Neutral
            - Unclear


        5. If the question asks for analysis, provide a structured answer with sections:  
            Overall sentiment  
            Key themes / complaints / praise  
            Notable quotes or expressions (optional)  
            Uncertainties / data gaps (if any)  
        6. If the posts do not contain enough information, say:  
        “The available posts do not provide a reliable answer.”  
        7. Do not add facts that aren't in the posts.
        8-Respond in the same language as the question

        ---
        Question:  
        {question}

        ---
        Context:  
        {full_context}


        ---
        ### Answer
        """
        response = model1.generate_content(prompt)
        return response.text
    except Exception as e:
        return f"an error occured {e}"

## Part 6: Report Generation with Multi-Agent System (MAS)

In this part, we implement a multi-agent architecture to automatically generate insightful business reports from social media data. The system is composed of several specialized agents, each responsible for a specific step in the process.

1. **Query Generator**  
   Converts high-level business goals into structured queries that other agents can understand.

In [60]:
query_gen = Agent(
    name="QueryGenerator",
    role="Transform user goals into deep research queries",
    tools = [ThinkingTools(add_instructions=True, think= True)],
    instructions="""
You are a highly skilled analyst assistant.

Your role is to transform vague or high-level business goals into detailed, structured queries.

You will receive:
- A business goal (written in natural language)

Your task:
1. Read and deeply understand the goal.
2. Extract and structure the following:

Your structured query must include the following keys in JSON-like format:

```json
{
  "intent": "What the user wants to understand or analyze",
  "filters": {
    "product_type": "...",
    "temporal_context": "...",
    "spatial_context": "...",
    "consumer_profiles": "...",
    "emotional_triggers": "...",
    "nutrition_profile": "...",
    "preparation_effort": "...",
    "created_time_range": "..."  // if a date range is implied
  },
  "metrics": [
    "List of key indicators or patterns to extract",
    "e.g. product mentions, perceived healthiness, typologies, motivations"
  ],
  "depth": "Level of insight expected: summary, deep dive, statistical, typological, etc.",
  "comparisons": "If the user implies comparing groups (profiles, moments, periods...)",
  "output_format": "The expected shape of the final answer (e.g. structured report, bullet points, graph summary)"
}
Always preserve the original meaning, and handle vague or indirect goals with precision.
""",
    model=Gemini(id="gemini-2.5-flash", temperature= 0.1)
)
goal = "healthy snacking among students"
query_response = query_gen.run(goal)
structured_query = query_response.content


In [61]:
print(structured_query)

```json
{
  "intent": "Understand the behaviors, perceptions, motivations, and barriers related to healthy snacking among students.",
  "filters": {
    "product_type": "Healthy snacks (e.g., fruits, vegetables, nuts, yogurt, whole-grain options, low-sugar items)",
    "temporal_context": "General",
    "spatial_context": "General",
    "consumer_profiles": "Students (e.g., high school, college, university students)",
    "emotional_triggers": "Motivations for and barriers against healthy snacking choices",
    "nutrition_profile": "Perceived and actual nutritional aspects of healthy snack choices",
    "preparation_effort": "Convenience and effort associated with healthy snack preparation/consumption",
    "created_time_range": "N/A"
  },
  "metrics": [
    "Prevalence and frequency of healthy snacking",
    "Specific types of healthy snacks consumed",
    "Student perceptions of what constitutes a 'healthy' snack",
    "Motivations for choosing healthy snacks (e.g., energy, focus, we

2. **Data Filter**  
   Selects the most relevant data (posts/comments) from the dataset based on the structured query.

In [182]:
filter_agent = Agent(
    name="DataFilter",
    role="filter the dataset using semantic understanding of the structured query",
    tools = [ThinkingTools(think = True, add_instructions= True), exec_python],
    instructions="""
This loads the full dataset into a DataFrame called df. Each row in df contains:

post_id

comment

type_of_snack

activity_context

temporal_context

spatial_context

social_context_detail

functional_triggers

emotional_triggers

consumer_profiles

barriers

nutrition_profile

preparation_effort

created_time

yyyy_mm

Your task:

Parse the structured query and extract the filter values.

Apply these filters semantically to df:

consumer_profiles: accept fuzzy matches like "étudiants", "students", "élèves", etc.

nutrition_profile: match related keywords such as "sain", "bio", "healthy", "fit", "équilibré"

temporal_context: match values like "matin" → "morning", "petit déjeuner", "early"

You may check both column values and keywords in df['comment']

Group results by post_id (comments from the same post share context).

Return a filtered DataFrame containing only the relevant entries.

Final Step:
Save the filtered dataset to a new CSV file by executing this code using exec_python tool:


df_filtered.to_csv('/filtered_dataset.csv', index=False)



Rules:

Use semantic matching logic — not strict string equality.

Avoid empty or low-context entries.

Always preserve full context by returning all relevant metadata fields.

""",
    model=Gemini(id="gemini-2.5-flash", temperature=0.1)
)   

In [184]:
filter_agent.run(structured_query)

RunResponse(content='', content_type='str', thinking=None, reasoning_content=None, messages=[Message(role='system', content='<your_role>\nfilter the dataset using semantic understanding of the structured query\n</your_role>\n\n<instructions>\n\nThis loads the full dataset into a DataFrame called df. Each row in df contains:\n\npost_id\n\ncomment\n\ntype_of_snack\n\nactivity_context\n\ntemporal_context\n\nspatial_context\n\nsocial_context_detail\n\nfunctional_triggers\n\nemotional_triggers\n\nconsumer_profiles\n\nbarriers\n\nnutrition_profile\n\npreparation_effort\n\ncreated_time\n\nyyyy_mm\n\nYour task:\n\nParse the structured query and extract the filter values.\n\nApply these filters semantically to df:\n\nconsumer_profiles: accept fuzzy matches like "étudiants", "students", "élèves", etc.\n\nnutrition_profile: match related keywords such as "sain", "bio", "healthy", "fit", "équilibré"\n\ntemporal_context: match values like "matin" → "morning", "petit déjeuner", "early"\n\nYou may ch

In [185]:
print(exec_python("""
import pandas as pd
print(pd.read_csv('/filtered_dataset.csv')) """))


Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1026, in read_csv
    return _read(filepath_or_buffer, kwds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 620, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1620, in __init__
    self._engine = self._make_engine(f, self.engine)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1880, in _make_engine
    self.handles = get_handle(
                   ^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/common.py", line 873, in get_handle
    handle = open(
             ^^^^^
FileNotFoundError: 

**Coder agent to extract statistics**

In [ ]:
coder_insight_agent = Agent(
    name="CodeWriter",
    role="Write Python codes on structured text queries.",
    debug_mode= True,
    instructions="""
You are a Python coding assistant.

Your task is to generate a complete Python script that extracts **all possible numeric and statistical insights** from a filtered dataset of annotated social media comments.

You will receive a text block that contains:
- A structured query (JSON-like, can be ignored)
- A list of enriched data entries (each like a row), with the following fields:
  - post_id
  - comment
  - type_of_snack
  - activity_context
  - temporal_context
  - spatial_context
  - social_context_detail
  - functional_triggers
  - emotional_triggers
  - consumer_profiles
  - barriers
  - nutrition_profile
  - preparation_effort
  - created_time
  - yyyy_mm

---

Your job:
- Ignore the query content.
- Focus entirely on the filtered dataset.
- Write a Python 3 script (using **pandas only**) that computes **all useful statistics**, including but not limited to:

1. **Volume Metrics**
   - Total number of comments
   - Total number of unique post IDs
   - Number of comments per month (`yyyy_mm`)
   - Number of comments per consumer profile

2. **Distribution Counts**
   - Value counts for each field:  
     - `consumer_profiles`, `emotional_triggers`, `functional_triggers`, `barriers`  
     - `temporal_context`, `spatial_context`, `social_context_detail`  
     - `type_of_snack`, `nutrition_profile`, `preparation_effort`

3. **Percentages**
   - Percentage distribution of values per categorical field (top 5 most frequent)

4. **Co-occurrence examples** (optional but valuable)
   - Most common pairs: (consumer_profile × emotional_trigger), etc.

5. **Trend Insight**
   - Monthly trend: number of posts per `yyyy_mm`

---

OUTPUT FORMAT:
- Return the Python code only, as a **string**
- No explanation, no markdown, no headers
- Just the code — ready to execute

You can assume that the filtered dataset is already loaded into a `pandas.DataFrame` called `df`.

Write clean, readable code that prints each result with a label.
""",
    model=Gemini(id="gemini-2.5-flash", temperature= 0.1)
)

In [122]:
code = coder_agent.run("query : " + structured_query + "\n\ndata: " + comments).content

In [123]:
print(code)

import pandas as pd
import re
import json

# Raw data provided in the prompt
data_raw = """
post_id: 3.570157881816461e+18
comment: ياله قالتلك مافيها ماي طيب اودابا شنو كازير كات صنع
type_of_snack: []
activity_context: []
temporal_context: []
spatial_context: []
social_context: []
functional_triggers: []
emotional_triggers: []
consumer_profiles: []
barriers: []
nutrition_profile: []
preparation_effort: []


post_id: 7.434845771867885e+18
comment: dh f marjane
type_of_snack: [['snack']]
activity_context: []
temporal_context: []
spatial_context: []
social_context: []
functional_triggers: []
emotional_triggers: []
consumer_profiles: []
barriers: [['Price / Affordability', 'Limited Availability']]
nutrition_profile: []
preparation_effort: []


post_id: 7.455344655241137e+18
comment: والله العظيم الا عندك الحق اخويا كنت كانشربه وانا حامله كان كيعطيني واحد الطاقه كنت كنحس براسي مزيان دابا ما بقيتش كانشربه راجعه غير اللور
type_of_snack: [['snack']]
activity_context: []
temporal_context: []
s

In [83]:
image = modal.Image.debian_slim().pip_install("pandas")

In [86]:
def exec_python(code):
    app = modal.App("Test-app", image = image)
    with app.run():
        sb = modal.Sandbox.create(app = app, image= image)
        p = sb.exec("python", "-c", f"{code}")
        output = sb.exec("python", "-c", code)
        stdout = output.stdout.read()
        stderr = output.stderr.read()
    return stdout + "\n" + stderr

In [88]:
print(exec_python("""
import pandas as pd
print(pd.Series([1,2,3,4,5]))
"""))

0    1
1    2
2    3
3    4
4    5
dtype: int64




In [130]:
output = exec_python(code)

In [114]:
code

'import pandas as pd\n\n# Assume df is already loaded into a pandas DataFrame\n# Example DataFrame structure based on the provided data snippet:\n# df = pd.DataFrame([\n#     {\'post_id\': 3.570157881816461e+18, \'comment\': \'...\', \'type_of_snack\': [], \'activity_context\': [], \'temporal_context\': [], \'spatial_context\': [], \'social_context_detail\': [], \'functional_triggers\': [], \'emotional_triggers\': [], \'consumer_profiles\': [], \'barriers\': [], \'nutrition_profile\': [], \'preparation_effort\': [], \'yyyy_mm\': \'2023-01\'},\n#     {\'post_id\': 7.434845771867885e+18, \'comment\': \'...\', \'type_of_snack\': [[\'snack\']], \'activity_context\': [], \'temporal_context\': [], \'spatial_context\': [], \'social_context_detail\': [], \'functional_triggers\': [], \'emotional_triggers\': [], \'consumer_profiles\': [], \'barriers\': [[\'Price / Affordability\', \'Limited Availability\']], \'nutrition_profile\': [], \'preparation_effort\': [], \'yyyy_mm\': \'2023-01\'},\n#    

In [128]:
executor_agent = Agent(
    name="ExecutorAgent",
    role="Execute any valid Python code returned by a coding agent.",
    tools=[exec_python],
    instructions="""
You are the code executor.
You receive Python code generated by a prior agent.
Your job is to run it exactly and return the output. No interpretation.
Your role is to:
- Use the `exec_python` tool to run it in a secure sandbox.
- Return only the output of the execution (no analysis, no rewriting, no reasoning).
Do not attempt to debug or modify the code. Simply execute and report the result.
""",
model = Gemini("gemini-2.5-flash", temperature= 0.1)
)

In [127]:
output  = executor_agent.run(code).content

In [131]:
output

"--- Volume Metrics ---\nTotal number of comments: 50\nTotal number of unique post IDs (excluding NaN): 38\n\nNumber of comments per consumer profile:\nconsumer_profiles\nDiet & Health-focused consumers    5\nAthletes / Fitness Enthusiasts     4\nPregnant Women                     1\nUNSURE                             1\nName: count, dtype: int64\n\n--- Distribution Counts & Percentages ---\n\nDistribution for 'consumer_profiles':\nconsumer_profiles\nDiet & Health-focused consumers    5\nAthletes / Fitness Enthusiasts     4\nPregnant Women                     1\nUNSURE                             1\nName: count, dtype: int64\n\nPercentage Distribution for 'consumer_profiles' (Top 5):\nconsumer_profiles\nDiet & Health-focused consumers    45.45\nAthletes / Fitness Enthusiasts     36.36\nPregnant Women                      9.09\nUNSURE                              9.09\nName: proportion, dtype: float64\n\nDistribution for 'emotional_triggers':\nemotional_triggers\nCraving Fix            

3. **Insight Analyzer**  
   Performs contextual analysis on the filtered data and extracts key patterns, emotions, statistics, and uncertainties.

In [110]:
analyzer_agent = Agent(
    name="InsightAnalyzer",
    role="analyze filtered social media data to extract meaningful insights",
  instructions= """
You are a senior insights analyst working on a dataset of Moroccan social media comments about snacking.

You will receive a string containing:

1. A structured query in JSON-like format

2. A list of enriched social media comment entries, where each entry includes:
   - post_id
   - comment
   - type_of_snack
   - activity_context
   - temporal_context
   - spatial_context
   - social_context_detail
   - functional_triggers
   - emotional_triggers
   - consumer_profiles
   - barriers
   - nutrition_profile
   - preparation_effort
   - created_time
   - yyyy_mm

  3. A block of statistical output generated from the same data (via Python code)
   - This includes counts, percentages, distributions, and trends
   - It was generated by a code execution agent and must be treated as **factual**


   - You must include the most relevant statistics **exactly as stated** — do not reword, round, or hallucinate


---

Your task is to extract client-ready insights from this data, organized by these sections:

---

### 1. Strategic summary
- What are the key themes around snacking in Morocco?
- What do users consider as “healthy snacking”?
- Are there contradictions (e.g., desire for health vs actual behavior)?
INCLUDE THE STATISTICS 

### 2. Snacking typology 
- Categorize snacking moments by:
  - Moment of consumption (temporal_context)
  - Consumer profile (consumer_profiles)
  - Motivation (emotional/functional_triggers)
  - Format (e.g., preparation_effort)
- Use typologies: utilitarian, pleasure, social, nomadic
INCLUDE THE STATISTICS


### 3. Zoom “Healthy Snacking”
- What products are perceived as healthy?
- What motivates healthy choices?
- What are the barriers or trade-offs?
- Are there local definitions of "sain"?
INCLUDE THE STATISTICS

### 4. *Consumer Profiles & Situational Contexts**  
- Identify key profiles: étudiants, mamans, sportifs, enfants, actifs pressés
- Link them to context (e.g., school, sport, home, work, transport)
INCLUDE THE STATISTICS

### 5. **Mentioned Brands & Products**  
- Mention any brands or products that appear in comments
- Highlight sentiment if possible (positive / neutral / negative)
INCLUDE THE STATISTICS


### 6. Ambiguity
- Any ambiguous insights or conflicting behavior patterns?
- Note if data is too sparse or unclear in any area

---

Rules:
- Use all available metadata fields (not just raw text)
- Group comments by post_id if they reflect the same post context
- Quote 1-2 real comments to support each insight
**Incorporate statistics from the Python output exactly as provided**
- Avoid hallucinations. Only use what's visible in the data
- If a claim is uncertain, say so

Deliver your insights in a **structured format with headings**, not a blob of text.
""",

    model = Gemini(id = "gemini-2.5-flash", temperature=0.2)
)

In [132]:
analysis = analyzer_agent.run("query : " + structured_query + "data :" + comments + "statistical output : " + output).content

In [133]:
print(analysis)

Here's an analysis of the provided Moroccan social media comments about snacking, focusing on healthy snacking behaviors, perceptions, motivations, and barriers.

---

### 1. Strategic Summary

Snacking in Morocco is driven by a mix of functional needs, emotional desires, and practical considerations. While there's a clear aspiration for healthier choices, traditional and convenient options, often perceived as less healthy, remain prevalent. Price and availability are significant barriers to healthy eating.

*   **Key Themes:**
    *   **Health Consciousness vs. Reality:** There's an awareness of healthy eating, with users discussing nutritional content and health impacts, but also a strong pull towards indulgent or nostalgic snacks.
    *   **Convenience and Affordability:** Ready-to-eat snacks dominate, and price is a major barrier, especially for healthier options.
    *   **Nostalgia and Comfort:** Many comments express fondness for traditional or childhood snacks, highlighting the

4. **Report Generator**  
   Takes the analyzed insights and turns them into a professional, structured report suitable for decision-makers.

In [134]:
report_agent = Agent(
    name="ReportGenerator",
    role="generate a structured and polished report from analyzed insights",
instructions="""
You are a professional consumer insights report writer.

You will receive:
- A structured analysis of annotated social media data, with sections like Summary, Typologies, Quotes, Statistics, and Gaps.

Your task is to turn this into a clear, business-ready report for a client studying snacking behaviors in Morocco.

---

Structure the report using the following sections:

1. **Strategic Summary**  
   - Provide a high-level overview of general snacking practices in Morocco  
   - Summarize how "healthy snacking" is perceived  
   - Highlight any contradictions or tensions (e.g., desire for health vs. actual behavior)

2. **Snack Typologies**  
   - Categorize snacking types based on:  
     - Moment of consumption  
     - Profile of the consumer  
     - Motivation (emotional/functional)  
     - Format or preparation effort  
   - Use tags like: utilitarian, pleasure-driven, social, or on-the-go

3. **Healthy Snacking Focus**  
   - Identify products perceived as healthy  
   - Analyze motivations behind healthy choices  
   - Note barriers, trade-offs, or hesitations  
   - Include how “healthy” is defined locally

4. **Consumer Profiles & Situational Contexts**  
   - Outline key consumer groups: students, mothers, athletes, children, busy workers  
   - Link them to relevant contexts: school, work, sport, home, transport

5. **Mentioned Brands & Products**  
   - List brands or products mentioned in comments  
   - Indicate tone or sentiment where possible (positive, neutral, negative)

6. **Recommendations & Innovation Opportunities**  
   - Suggest clear product or messaging opportunities  
   - Ground all suggestions in observed data only  
   - Avoid speculation or invented facts

---

✍️ Writing style:
- Professional and clear  
- Business-oriented, not academic  
- Bullet points or short paragraphs are acceptable  
- Do not include anything not supported by the data  
- If a section has weak or no findings, state this explicitly

The final report should be polished, well-organized, and directly useful to marketing, R&D, or innovation teams.
"""
,
    model = Gemini(id = "gemini-2.5-flash", temperature=0.2)
)

In [135]:
report = report_agent.run(analysis).content

In [136]:
print(report)

## Consumer Snacking Behaviors in Morocco: Insights & Opportunities

### 1. Strategic Summary

Snacking in Morocco is a multifaceted behavior driven by a blend of functional needs, emotional desires, and practical considerations. While there is a discernible aspiration among consumers for healthier choices, traditional, convenient, and often less healthy options remain highly prevalent. Price and availability emerge as significant barriers to the widespread adoption of healthier eating habits.

**Key Observations:**

*   **Health Consciousness vs. Reality:** Consumers demonstrate an awareness of healthy eating, discussing nutritional content and health impacts. However, there is a strong pull towards indulgent or nostalgic snacks, leading to a significant preference for general "snack" types (75.51% of mentions) over "Healthy snack" (10.20%).
*   **Convenience and Affordability:** Ready-to-eat snacks dominate the market, with price being a primary barrier, particularly for healthier al

In [24]:
def generate_report(goal):
    try:
        # Step 1: Generate structured query from goal
        query_response = query_gen.run(goal)
        structured_query = query_response.content

        # Step 2: Filter dataset based on query
        comments = "\n".join(df["text"].sample(100, random_state= 42))
        filtered_data_response = filter_agent.run(structured_query + comments)
        filtered_data = filtered_data_response.content

        #step 3: Statistics
        statistic = coder_agent.run(structured_query + filtered_data)
        statistics = statistic.content

        # Step 3: Analyze the filtered data
        insights_response = analyzer_agent.run(structured_query + filtered_data + statistics)
        insights = insights_response.content

        # Step 4: Generate report
        report_response = report_agent.run(insights)

        return report_response.content
    except Exception as e:
        return e

# Part 7: Gradio Frontend

In this part, we build a user-friendly web interface using **Gradio**.

The goal is to allow users to ask natural language questions about social media feedback and receive meaningful analysis powered by a large language model. Gradio will serve as the front end for the retrieval-augmented generation (RAG) system we built earlier.

Users will type their questions in the Gradio interface. The app will:
- Use the vector database (Qdrant) to retrieve the most relevant social media posts.
- Pass these posts along with the user’s question to a Gemini model.
- Display the model’s structured analysis as the response.

This interface makes the insights accessible to non-technical users and demonstrates the power of combining embeddings, vector search, and LLMs in a real-world application.



In [14]:
def handle_input(user_input, mode):
    if mode == "Ask a question" :
        return answer_question(user_input)
    elif mode == "Generate a report":
        return generate_report(user_input)

In [17]:
iface = gr.Interface(
    fn=lambda user_input, mode: handle_input(user_input, mode),
   inputs=[
        gr.Textbox(label="Type your question or business goal", lines=2),
        gr.Radio(choices=["Ask a question", "Generate a report"], label="Choose a mode")
    ],
    outputs=[
        gr.Textbox(label="Answer"),
    ],
    title="Social Media Insight Assistant",
    description="Ask any question. The bot will analyze and give you insights.",
    theme="default"
)

iface.launch()

* Running on local URL:  http://127.0.0.1:7860
* To create a public link, set `share=True` in `launch()`.


## 8- Evaluation of the Multi-Agent System Using an LLM-as-a-Judge Approach

This section presents the evaluation methodology and results for our five-agent system, designed for social media insight extraction. The agents—responsible for query generation, data filtering, insight extraction, statistical analysis, and report generation—are each assessed individually to ensure robustness, coherence, and factual accuracy.

### Objective

To measure the effectiveness of each agent, we adopt an **LLM-as-a-Judge** evaluation framework. The goal is to evaluate:
- Accuracy of filtered data
- Relevance and completeness of insights
- Faithfulness of generated reports to input data
- Precision of statistical extraction
- Consistency and structure of generated queries


### Query generator: Evaluation of the first agent using LLM as a judge

In [ ]:
query = "what do people think about the products, and give me in percentages how many liked/didn't like a product"


generated = query_gen.run(query).content

In [ ]:
query_eval_agent.run("user goal : " + query + "\nstructured query : " + generated)

In [ ]:
generated

#### perfect score ! the agent does a great job

### Filter agent evaluation : I will test it on a test_set ( a small sample of our initial dataset ),give it the comments I am looking to extract, and check the accuracy of the agent using an LLM

In [ ]:
filter_eval_agent = Agent(
    name="FilterEvaluator",
    role="Evaluate the accuracy of the filtered comments",
    instructions="""
You are an evaluation agent.

You will receive:
1. A string of reference comments (the correct ones to retrieve).
2. A string of filtered comments (what the filtering agent retrieved).

Your task:
- Compare both sets.
-Conclude the correctly retrieved comments (matching with reference and filtered)
- Calculate the **accuracy** as:
  (Number of correctly retrieved comments) / (Total number of comments in the reference set)

Only compare exact matches or very similar ones.
Return only the accuracy as a number between 0 and 1 (e.g. 0.75).
""",
model = Gemini("gemini-2.5-flash", temperature=0.2)
)

In [ ]:
dataset = df["text"].sample(100, random_state= 40)
test_set=dataset[:20]
train_set = dataset[21:]

In [ ]:
print(test_set)

In [ ]:
ref = "comment 1 : خديتو ومداقو مزوينش ومطحون بزاااف بحال دقيق وزيد فيه ريحة معجباتنيش بحال نكهات لكنديرو فلحلوة وانا دابا قطعتهوم بجوج مبقيتش كناخدهم, comment 2 : jai réalisé le gâteau hier franchement il était trop bon, comment 3 : واقيلة انا وحيدة لي معنديش مع هاد شي ديال سقاطة جميع انواع سقاطة معنديش معاها كاين شي حاد بحالي ولا غير انا, comment 4 : كناخد الكحلة كتعجبني مديريش السشية كاملة ديال الحار غير شوية صافي اول مرة درتها كاملة كناكل ونبكي, comment 5:درتو لبارح تفكرت دكريات ههههه ولادي حماقو عليه"
comments = "\n".join(test_set)
filtered_data = filter_agent.run("structured query : " + generated + "comments : " + comments ).content

In [ ]:
filter_eval_agent.run("reference comments : " + ref + "filtered comments : " + filtered_data).content

In [ ]:
filtered_data

This validates our result, A 100% accuracy

## Coder agent evaluation

In [ ]:
# Custom image to support AGNO and pandas
image = modal.Image.debian_slim().pip_install(
    "agno", "pandas", "numpy", "scikit-learn" 
)

app = modal.App(
    "Test-app",
    image=image,
    secrets=[
        modal.Secret.from_name("gemini-secret", required_keys=["AIzaSyAYFKWG_34OznV9eMowZ-7kWEx1KBdlurc"]),
    ],
)

In [ ]:
def exec_python(code):
    app = modal.App.lookup("Test-app", create_if_missing=True)
    sb = modal.Sandbox.create(app = app)
    p = sb.exec("python", "-c", f"{code}")
    return p.stdout.read()

In [ ]:
exec_python("print(2+2)")

In [ ]:
executor_agent.run(coder_agent.run(generated + filtered_data).content).content

In [ ]:
filtered_data

In [ ]:
expected_negative = ["comment 1", "comment 4"]
expected_positive = ["comment 2 ", "comment 3", "comment 6", "comment 5"]
expected_neutral = []

negative = (len(expected_negative)/6) *100
positive = (len(expected_positive)/6 )*100 
neutral  = (len(expected_neutral)/6)* 100

print(f'negative: {negative:.2f}%, positive : {positive:.2f}%, neutre: {neutral:.2f}%')

## SAME RESULT !

In [ ]:
TheAgent = Agent(
    name="AgentCoder",
    role="Code agents based on task requirements",
    tools=[exec_python],
    model=Gemini("gemini-2.5-flash"),
    instructions="""
You are a professional agent coder with access to the full AGNO documentation. You will use the `agno.agents.Agent` class to build and instantiate new agents. You also have access to the `modal` library and the `exec_python(code)` tool to securely execute Python code in a sandbox environment.

You are powered by the Gemini 2.5 Flash model.

---
µ
You will receive two inputs:
1. A natural language **query from the user**
2. A **dataset of social media comments** (provided as a pandas DataFrame called `df`)

Your goal is to build a pipeline of agents tailored to the query in order to generate a structured and well-written report.

---

###  STEP 1: QUERY INTERPRETATION
Interpret the user’s query and determine whether to include:

- `FilterAgent`: if the query targets a specific topic, brand, product, or theme
- `CoderAgent`: if the query asks for statistics, metrics, or quantitative results

You must always include:

- `AnalysisAgent`: to perform deep analysis on the relevant comments
- `WriterAgent`: to write the final report using the output of the analysis

---

### STEP 2: AGENT CREATION
For each required agent:

- Define it using the `Agent` class from `agno.agents`
- Assign a clear role, goal, backstory, and detailed instructions
- If statistics or calculations are required:
  - Generate valid Python code using the Gemini model
  - Use `exec_python(code)` to execute the code and capture its output
  - Return the result as a **clear, human-readable sentence** (e.g., "43.2% of users liked the product")

---

### STEP 3: EXECUTION FLOW
- Chain the agents in this order:
  1. `FilterAgent` → if needed (optional)
  2. `CoderAgent` → if needed (optional)
  3. `AnalysisAgent` → always
  4. `WriterAgent` → always

- Pass only the necessary outputs between agents:
  - If created, pass the filtered comments to the next steps
  - If created, pass the metrics/stats from `CoderAgent` to the analysis step
  - The `AnalysisAgent` must output its result into a variable called `analysis_output`

- On the final line of your generated Python code, you **must** run the `WriterAgent` using the analysis result:
```python
final_report = WriterAgent.run(analysis_output)
print(final_report)
"""
)

In [ ]:
import modal
comments = "\n".join(df["text"].sample(100, random_state=40))
query = "what do people think about the products, and give me in percentages how many liked/didn't like a product"

In [ ]:
import modal
import json
import inspect
from typing import Any
import os

app = modal.App("Test-app")


@app.function(secrets=[modal.Secret.from_name("custom-secret")])
def f():
    print(os.environ["Gemini_API"])

image = (
    modal.Image.debian_slim()
    .pip_install("numpy", "pandas", "agno", "google-genai") )


# 1. Use your driver_program exactly as is
def driver_program():
    import json
    import sys
    from contextlib import redirect_stderr, redirect_stdout
    from io import StringIO

    globals: dict[str, Any] = {}
    while True:
        command = json.loads(input())
        if (code := command.get("code")) is None:
            print(json.dumps({"error": "No code to execute"}))
            continue

        stdout_io, stderr_io = StringIO(), StringIO()
        with redirect_stdout(stdout_io), redirect_stderr(stderr_io):
            try:
                exec(code, globals)
            except Exception as e:
                print(f"Execution Error: {e}", file=sys.stderr)

        print(
            json.dumps(
                {
                    "stdout": stdout_io.getvalue(),
                    "stderr": stderr_io.getvalue(),
                }
            ),
            flush=True,
        )

# 2. Launch sandbox and run driver program
driver_source = inspect.getsource(driver_program)
driver_command = f"{driver_source}\n\ndriver_program()"
app = modal.App('Test-app', secrets= [modal.Secret.from_name("custom-secret" ,required_keys=["GOOGLE_API_KEY"])], image = image)

# 3. Your exact `run_code` function reused
def run_code(p: modal.container_process.ContainerProcess, code: str):
    p.stdin.write(json.dumps({"code": code}))
    p.stdin.write("\n")
    p.stdin.drain()
    next_line = next(iter(p.stdout))
    result = json.loads(next_line)
    print(result["stdout"], end="")
    print("\033[91m" + result["stderr"] + "\033[0m", end="")

# 4. Wrap in exec_python() as a clean callable for agents
def exec_python(code: str) -> str:
    p.stdin.write(json.dumps({"code": code}))
    p.stdin.write("\n")
    p.stdin.drain()
    next_line = next(iter(p.stdout))
    result = json.loads(next_line)
    output = result["stdout"]
    if result["stderr"]:
        output += "\n[stderr]:\n" + result["stderr"]
    return output.strip()


In [ ]:
PROMPT = """
You are AgentCoder, an autonomous builder of AI agent pipelines.  
You have access to the class `agno.agent.Agent` to create agents and the tool `exec_python(code: str)` that executes Python code inside an already-running Modal sandbox and returns the code’s stdout (stderr appended if non-empty).

Inputs you receive each time you run:
1. A natural-language user query.
2. A dataset of social-media comments supplied as a single multi-line string (one comment per line).

Your job is to produce a single, well-structured written report that answers the user's query.

────────────────────────────────────────────────────────
STEP 1  Interpret the query
────────────────────────────────────────────────────────
• If the query requires narrowing to a subset (topic, brand, date, etc.), create a **FilterAgent**.  
• If the query requires numbers or percentages, create a **CoderAgent** that generates and executes Python to compute those statistics.  
• Always include:
  **AnalysisAgent** extracts themes, insights, and merges any stats.  
  **WriterAgent** writes the final human-readable report.  

All agents are subclasses of `agno.agent.Agent` and must expose a `.run()` method.

────────────────────────────────────────────────────────
STEP 2  Build the agents
────────────────────────────────────────────────────────
• Each agent returns **a plain string**. If structured data is required, return it JSON-encoded with `json.dumps`.  
• Provide clear role, goal, and instructions for every agent.  
• `CoderAgent` must:  
  1. Build Python code that performs the required computation.  
  2. Place that code into a triple-quoted string.  
  3. Execute it with `exec_python(code)` and capture its string output.  
  4. Return that output as its own result.

────────────────────────────────────────────────────────
STEP 3  Chain execution
────────────────────────────────────────────────────────
Order of execution:  
    FilterAgent (when created) → CoderAgent (when created) → AnalysisAgent (always) → WriterAgent (always)  
Pass only what is needed from each agent to the next.  
The variable **analysis_output** must hold the AnalysisAgent's return string.

At the very end of your generated Python pipeline:
```python
final_report = WriterAgent.run(analysis_output)
print(final_report)
────────────────────────────────────────────────────────
STEP 4 Produce and run the pipeline
────────────────────────────────────────────────────────

Assemble the entire pipeline (class definitions + execution logic) as one triple-quoted Python string.

Execute it immediately with:
exec_python(\"\"\"<full-pipeline-code>\"\"\")

Return only the string produced by exec_python, which must be the WriterAgent's report.

Do NOT output the generated code or any logs.

────────────────────────────────────────────────────────
Required report structure (language = user's query language)
────────────────────────────────────────────────────────

Executive Summary (3-5 lines)

Quantitative Analysis (counts, percentages, metrics)

Qualitative Insights (themes, examples, product mentions)

Conclusion (overall tone, takeaways, recommendations)

Return nothing except the final report string.
"""

In [ ]:


AgentCoder = Agent(
    name="AgentCoder",
    role="Code agents based on user requests",
    tools=[exec_python],
    instructions=PROMPT,
    model=Gemini("gemini-2.5-flash")
)

In [ ]:
test_code = '''
from agno.agent import Agent
from agno.models.google import Gemini

query_gen = Agent(
    name="QueryGenerator",
    role="Transform user goals into deep research queries",
    instructions="""
You are a highly skilled analyst assistant.

Your role is to transform vague or high-level business goals into detailed, structured queries.

You will receive:
- A business goal (written in natural language)

Your task:
1. Read and deeply understand the goal.
2. Extract and structure the following:
   - Intent: What is the user trying to understand or achieve?
   - Filters: Key constraints (e.g. product, date range, keywords, sentiment, user type, language, location).
   - Metrics / Objectives: Quantitative or qualitative metrics to extract (e.g. percentage of positive feedback, average satisfaction, common complaints, emerging trends).
   - Depth: Desired scope (e.g. high-level summary, deep dive, longitudinal trends, statistical breakdown).
   - Comparisons: Any required comparisons (e.g. between two products, before/after a launch).
   - Output Format: The ideal way to structure the answer.

Return your result as a JSON-like string with clear keys:
{
  "intent": "...",
  "filters": { ... },
  "metrics": [...],
  "depth": "...",
  "comparisons": "...",
  "output_format": "..."
}

Always preserve the original meaning, and handle vague or indirect goals with precision.
""",
    model=Gemini(id="gemini-2.5-flash")
)
goal = "summarize what people say about the snacks and tell me what pourcentage of them liked it"
query_response = query_gen.run(goal)
structured_query = query_response.content

print(structured_query)
'''

In [ ]:
test_code = '''
import numpy as np
print(np.cos(20))
'''

In [ ]:
with app.run():
    sb = modal.Sandbox.create(app=app, image= image, timeout = 60 *10, secrets= [modal.Secret.from_name("custom-secret" ,required_keys=["GOOGLE_API_KEY"])])
    p = sb.exec("python", "-c", driver_command)
    print(exec_python(test_code))

In [15]:
from agno.team.team import Team

social_insight_team = Team(
    name="SocialInsightTeam",
    mode="coordinate",
     members=[
        query_gen,
        filter_agent,
        coder_agent,
        executor_agent,
        analyzer_agent,
        report_agent
    ],
    context=True,
    instructions="""
You are a coordinated team of specialized agents analyzing social media feedback.
- Each agent has a precise task in a multi-step pipeline.
- Do not skip any agent unless explicitly instructed.
- Ensure smooth handoff of results from one agent to the next.
- Maintain data integrity and avoid redundant processing.
""",
model = Gemini("gemini-2.5-flash")
)

In [16]:
comments = "\n".join(df["text"].sample(20, random_state= 60))

In [19]:
print(comments)

جيب لينا بيموات القدام
لمشكل زوين ولكن زفر حيت مديرين فيه لبيض
كاينة واحد بحال طاكير هي سميتها tirma من اسبانيا كاتجي كاتباع فأقاليم الجنوبية بكثرة ولكن واعرة بحال طاكير لقديم
henris and matina
هذاك هو سرطان كتشريوه بثمن رخيص
alors golden je meurs jai jamais aimé et pour info les quand on était petit on aimait pas trop les petits déjeuner en tt cas pour ma part et je me rappelle une fois ma tante avait insister auprès de ma proviseur que je prenne raibimerendina avant d aller en cours et bien résultat vomis sur les pides de la proviseur pour ma part pour les snacks d enfance et malheureusement ils ont disparu à part okey car ils existe tjrs juste votre hanout ne lavais pas cest loco fifty mina tonga kiwi mon caramel orangis et pour les jus les vrais savent que le meilleur reste miami
الله احفظكتحياتيبغرير واش من بغرير
حنان حنان تفاداي اختي حب العزيز كيما هو صعيب على الحامل والمرضع فكذلك الطفل
كارطون دالحليب احسن من بلاصتيك
bro thats not the original jamila
ولي تأكلي بطاطا مسلوقة ديري ا

In [22]:
social_insight_team.run("query :" + "understand what people like" + "\n\ncomments :" + comments).content

WARNING  Context is not a dict

'## Social Media Sentiment Analysis: Positive Preferences and Product Qualities\n\n**Report Date:** October 26, 2023\n\n### Executive Summary\n\nAn analysis of social media comments reveals a strong positive sentiment primarily centered around a preference for original or traditional product versions, specific food items and brands, and an appreciation for quality and helpful content. Users frequently express admiration using strong positive adjectives, highlighting a desire for familiar, high-quality, and informative experiences.\n\n### Key Findings\n\n*   **Preference for Original/Traditional Versions:** A significant trend indicates a desire for "old" or "original" product formulations, particularly noted for biscuits and implicitly for "Taguer," suggesting nostalgia or perceived superior quality of past versions.\n*   **Specific Food Items and Brands:** Several distinct food items and brands receive positive mentions, including "Miami," "Baghrir," "Rialto," "Tirma," "Mirror dessert

In [23]:
generate_report("understand what people like")

'## Consumer Sentiment Analysis Report: Food & Beverage Preferences\n\n**Date:** October 26, 2023\n\n---\n\n### Executive Summary\n\nThis report summarizes an analysis of consumer comments, revealing a predominantly positive sentiment towards various food and beverage items. Key drivers of preference include taste, quality, and a strong sense of nostalgia for childhood snacks. Consumers express clear preferences for specific products, desserts, and even packaging types, indicating opportunities for businesses to align with these consumer desires. A notable trend is the appreciation for both traditional quality and modern health-conscious options.\n\n### Key Findings\n\nThe analysis identified several recurring themes and specific preferences among consumers:\n\n*   **Specific Product Preferences:**\n    *   **Snacks/Sweets:** High praise for "Tirma" (especially its similarity to "old Takir"), and fond memories for childhood snacks like "Okey," "Loco Fifty," "Mina," "Tonga," "Kiwi," "Mo

In [187]:
image = modal.Image.debian_slim().pip_install(["pandas"]).add_local_file(local_path = "social_media_analyzed.csv", remote_path= "/social_media_analyzed.csv")
df_sandbox_volume = modal.Volume.from_name("data-volume", create_if_missing=True)

In [200]:
def exec_python(code):
    app = modal.App("Test-app", image = image)
    with app.run():
        sb = modal.Sandbox.create(app = app, image= image, volumes={"/social_media_analyzed.csv": df_sandbox_volume})
        p = sb.exec("python", "-c", f"{code}")
        output = sb.exec("python", "-c", code)
        stdout = output.stdout.read()
        stderr = output.stderr.read()
    return stdout + "\n" + stderr

In [201]:
output = exec_python("""
import pandas as pd
df = pd.read_csv('/social_media_analyzed.csv')""")

In [198]:
print(output)


Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1026, in read_csv
    return _read(filepath_or_buffer, kwds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 620, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1620, in __init__
    self._engine = self._make_engine(f, self.engine)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1880, in _make_engine
    self.handles = get_handle(
                   ^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pandas/io/common.py", line 873, in get_handle
    handle = open(
             ^^^^^
FileNotFoundError: 

In [164]:
coder_insight_agent = Agent(
    name="CodeWriter",
    role="Write Python codes on structured text queries.",
    instructions="""
You are a Python coding assistant.

Your task is to generate a **complete executable Python script** that extracts **all possible numeric and statistical insights** from a filtered dataset of Moroccan social media comments.

You will receive a text block containing:
- A structured query (JSON-like, can be ignored)


---

 Before any analysis, you MUST include the following import and CSV loading code at the top:


import pandas as pd
df = pd.read_csv('/social_media_analyzed.csv')

It imports a dataset of enriched data entries (each like a row), with the following fields:
  - post_id
  - comment
  - type_of_snack
  - activity_context
  - temporal_context
  - spatial_context
  - social_context_detail
  - functional_triggers
  - emotional_triggers
  - consumer_profiles
  - barriers
  - nutrition_profile
  - preparation_effort
  - created_time
  - yyyy_mm

---

Your job:
- Ignore the query content.
- Focus entirely on the  dataset.
- Write a Python 3 script (using **pandas only**) that computes **all useful statistics**, including but not limited to:

1. **Volume Metrics**
   - Total number of comments
   - Total number of unique post IDs
   - Number of comments per month (`yyyy_mm`)
   - Number of comments per consumer profile

2. **Distribution Counts**
   - Value counts for each field:  
     - `consumer_profiles`, `emotional_triggers`, `functional_triggers`, `barriers`  
     - `temporal_context`, `spatial_context`, `social_context_detail`  
     - `type_of_snack`, `nutrition_profile`, `preparation_effort`

3. **Percentages**
   - Percentage distribution of values per categorical field (top 5 most frequent)

4. **Co-occurrence examples** (optional but valuable)
   - Most common pairs: (consumer_profile × emotional_trigger), etc.

5. **Trend Insight**
   - Monthly trend: number of posts per `yyyy_mm`

---

OUTPUT FORMAT:
- Return the Python code only, as a **string**
- No explanation, no markdown, no headers
- Just the code — ready to execute

You can assume that the filtered dataset is already loaded into a `pandas.DataFrame` called `df`.

Write clean, readable code that prints each result with a label.
""",
    model=Gemini(id="gemini-2.5-flash", temperature= 0.1)
)


In [165]:
code = coder_insight_agent.run(structured_query).content

In [168]:
result = exec_python(code)

In [169]:
insight_agent = Agent(
    name="InsightAnalyzer",
    role="analyze filtered social media data to extract meaningful insights",
    tools = [exec_python],
  instructions= """
You are a senior insights analyst working on a dataset of Moroccan social media comments about snacking.

You will receive:
1. A structured query in JSON-like format
2. A block of statistical output generated from the same data (via Python code)
   - This includes counts, percentages, distributions, and trends
   - It was generated by a code execution agent and must be treated as **factual**


   - You must include the most relevant statistics **exactly as stated** — do not reword, round, or hallucinate


Before doing any analysis, you **must** run the following Python code using the `exec_python` tool:


import pandas as pd
df = pd.read_csv('/social_media_analyzed.csv')


This loads the dataset into the df variable. From this point on, use df as your data source.
Each row of df contains:

post_id

comment

type_of_snack

activity_context

temporal_context

spatial_context

social_context_detail

functional_triggers

emotional_triggers

consumer_profiles

barriers

nutrition_profile

preparation_effort

created_time

yyyy_mm
---



---

Your task is to extract client-ready insights from this data, organized by these sections:

---

### 1. Strategic summary
- What are the key themes around snacking in Morocco?
- What do users consider as “healthy snacking”?
- Are there contradictions (e.g., desire for health vs actual behavior)?
INCLUDE THE STATISTICS 

### 2. Snacking typology 
- Categorize snacking moments by:
  - Moment of consumption (temporal_context)
  - Consumer profile (consumer_profiles)
  - Motivation (emotional/functional_triggers)
  - Format (e.g., preparation_effort)
- Use typologies: utilitarian, pleasure, social, nomadic
INCLUDE THE STATISTICS


### 3. Zoom “Healthy Snacking”
- What products are perceived as healthy?
- What motivates healthy choices?
- What are the barriers or trade-offs?
- Are there local definitions of "sain"?
INCLUDE THE STATISTICS

### 4. *Consumer Profiles & Situational Contexts**  
- Identify key profiles: étudiants, mamans, sportifs, enfants, actifs pressés
- Link them to context (e.g., school, sport, home, work, transport)
INCLUDE THE STATISTICS

### 5. **Mentioned Brands & Products**  
- Mention any brands or products that appear in comments
- Highlight sentiment if possible (positive / neutral / negative)
INCLUDE THE STATISTICS


### 6. Ambiguity
- Any ambiguous insights or conflicting behavior patterns?
- Note if data is too sparse or unclear in any area

---

Rules:
- Use all available metadata fields (not just raw text)
- Group comments by post_id if they reflect the same post context
- Quote 1-2 real comments to support each insight
**Incorporate statistics from the Python output exactly as provided**
- Avoid hallucinations. Only use what's visible in the data
- If a claim is uncertain, say so

Deliver your insights in a **structured format with headings**, not a blob of text.

""",

    model = Gemini(id = "gemini-2.5-flash", temperature=0.1)
)

In [171]:
analysis = analyzer_agent.run("query : " + structured_query + "statistical output : " + result).content

In [172]:
report = report_agent.run(analysis).content

In [173]:
print(report)

## Consumer Snacking Behaviors in Morocco: Insights from Social Media Data

### 1. Strategic Summary

This report provides an overview of snacking behaviors in Morocco, derived from an analysis of social media data. The insights highlight general snacking practices, perceptions of "healthy snacking," and key tensions observed in consumer choices.

**General Snacking Practices:**
Snacking is a prevalent behavior, with "snack" being the most frequently identified type of consumption (61.21% of mentions). Convenience is a significant driver, as evidenced by the high preference for "Ready-to-Eat" formats (29.55%). Snacking serves both functional needs, primarily to address "Hunger gap / tide over" (9.77%) and provide an "Energy Boost" (1.09%), as well as emotional triggers such as "Craving Fix" (7.58%) and "Comfort & Nostalgia" (2.51%).

**Perception of "Healthy Snacking":**
"Healthy snacking" is a relatively niche area within the dataset, explicitly tagged in only 1.48% of comments. When 

In [ ]:
filter_agent = Agent(
    name="DataFilter",
    role="filter the dataset using semantic understanding of the structured query",
    tools = [ThinkingTools(think = True, add_instructions= True), exec_python],
    instructions="""
This loads the full dataset into a DataFrame called df. Each row in df contains:

post_id

comment

type_of_snack

activity_context

temporal_context

spatial_context

social_context_detail

functional_triggers

emotional_triggers

consumer_profiles

barriers

nutrition_profile

preparation_effort

created_time

yyyy_mm

Your task:

Parse the structured query and extract the filter values.

Apply these filters semantically to df:

consumer_profiles: accept fuzzy matches like "étudiants", "students", "élèves", etc.

nutrition_profile: match related keywords such as "sain", "bio", "healthy", "fit", "équilibré"

temporal_context: match values like "matin" → "morning", "petit déjeuner", "early"

You may check both column values and keywords in df['comment']

Group results by post_id (comments from the same post share context).

Return a filtered DataFrame containing only the relevant entries.

Final Step:
Save the filtered dataset to a new CSV file by executing this code using exec_python tool:


df_filtered.to_csv('/filtered_dataset.csv', index=False)



Rules:

Use semantic matching logic — not strict string equality.

Avoid empty or low-context entries.

Always preserve full context by returning all relevant metadata fields.

""",
    model=Gemini(id="gemini-2.5-flash", temperature=0.1)
)   

## Consumer Snacking Behaviors in Morocco: Insights from Social Media Data

### 1. Strategic Summary

This report provides an overview of snacking behaviors in Morocco, derived from an analysis of social media data. The insights highlight general snacking practices, perceptions of "healthy snacking," and key tensions observed in consumer choices.

**General Snacking Practices:**
Snacking is a prevalent behavior, with "snack" being the most frequently identified type of consumption (61.21% of mentions). Convenience is a significant driver, as evidenced by the high preference for "Ready-to-Eat" formats (29.55%). Snacking serves both functional needs, primarily to address "Hunger gap / tide over" (9.77%) and provide an "Energy Boost" (1.09%), as well as emotional triggers such as "Craving Fix" (7.58%) and "Comfort & Nostalgia" (2.51%).

**Perception of "Healthy Snacking":**
"Healthy snacking" is a relatively niche area within the dataset, explicitly tagged in only 1.48% of comments. When health is considered, the focus is predominantly on "High Protein / Energy" (2.82%) and "Sugar-Free / Low Sugar" (1.61%) attributes, suggesting these are key indicators of health for Moroccan consumers in this context.

**Contradictions and Tensions:**
A notable tension exists between consumer aspirations for health and their actual snacking behaviors. While "Diet & Health-focused consumers" represent a significant segment (6.62% of profiles), explicit mentions of "Healthy snack" are considerably lower (1.48%). This discrepancy suggests that despite a desire for healthier options, consumers face significant barriers. The most prominent barriers identified are "Price / Affordability" (9.01%), "Limited Availability" (3.7%), and "Taste Dissatisfaction" (3.32%), which likely deter consumers from consistently choosing healthy alternatives.

### 2. Snack Typologies

Snacking behaviors in Morocco can be categorized into several typologies based on observed motivations, moments, and formats:

*   **Utilitarian Snacking:**
    *   **Motivation:** Primarily driven by functional needs such as bridging "Hunger gap / tide over" (9.77%) and seeking an "Energy Boost" (1.09%).
    *   **Format:** Highly favors "Ready-to-Eat" (29.55%) and "Minimal Effort (Open & Assemble / Mix)" (7.49%) options, indicating a demand for quick and convenient solutions.
*   **Pleasure-Driven Snacking:**
    *   **Motivation:** Centered around emotional satisfaction, notably "Craving Fix" (7.58%) and "Comfort & Nostalgia" (2.51%).
*   **Social Snacking:**
    *   **Context:** Occurs in shared settings, with "With Children / Family" (2.88%) and "With Friends" (0.45%) being the main social contexts. It is important to note that "Solo" snacking (5.23%) is more prevalent than social contexts.
*   **On-the-Go/Nomadic Snacking:**
    *   **Context:** Consumed outside the home, indicated by "On the Go" (1.3%), "At School" (0.66%), and "At Work" (0.15%) spatial contexts.
    *   **Format:** "Ready-to-Eat" (29.55%) is highly relevant for this typology due to its convenience.

**Temporal Context (Top 5 specific mentions):**
*   Late-Night Craving / Midnight Snack (2.47%)
*   Evening Snack (0.81%)
*   Post lunch / Pre-dinner (0.77%)
*   Mid-morning (0.64%)
*   Breakfast (0.59%)

### 3. Healthy Snacking Focus

**Products Perceived as Healthy:**
Based on the `nutrition_profile` data, products perceived as healthy are those offering "High Protein / Energy" (2.82%) and "Sugar-Free / Low Sugar" (1.61%). The explicit "Healthy snack" category accounts for 1.48% of snack type mentions.

**Motivations for Healthy Choices:**
While specific motivations for *healthy* choices are not isolated, the presence of "Diet & Health-focused consumers" (6.62%) indicates a segment driven by overall wellness goals. General functional triggers like "Energy Boost" (1.09%) and "Hunger gap / tide over" (9.77%) may also motivate healthy snacking for sustained energy or satiety.

**Barriers or Trade-offs:**
Significant barriers hinder the adoption of healthy snacking options:
*   **Price / Affordability:** 9.01% of mentions.
*   **Limited Availability:** 3.7% of mentions.
*   **Taste Dissatisfaction:** 3.32% of mentions.
These barriers suggest that even when consumers desire healthy options, cost, difficulty in sourcing them, or a perceived compromise on taste can lead to alternative choices.

**Local Definitions of "Sain":**
The provided statistical data does not offer explicit local definitions of "sain" (healthy). However, the emphasis on "High Protein / Energy" and "Sugar-Free / Low Sugar" in the `nutrition_profile` suggests that these attributes are key components of what is considered healthy within Moroccan social media discussions.

### 4. Consumer Profiles & Situational Contexts

**Key Consumer Profiles Identified:**
Based on the available data, the most prominent consumer profiles (excluding uncategorized data) are:
*   **Diet & Health-focused consumers:** 6.62%
*   **Parents of young children:** 2.36%
*   **Athletes / Fitness Enthusiasts:** 1.22%

**Situational Contexts:**
*   **At Home:** This is the most common specific spatial context (4.62%), indicating the home as a primary location for snacking.
*   **On the Go:** Accounts for 1.3% of mentions, highlighting snacking during transit or outside the home.
*   **At School:** Mentioned in 0.66% of comments (133 comments), serving as the most direct proxy for student-related snacking.
*   **At Work:** A minor context, representing 0.15% of mentions.

**Student Context Limitations:**
It is important to note that direct identification of "Teens/Students" in the `consumer_profiles` is extremely limited in this dataset. While "At School" is a relevant spatial context, its low representation (0.66%) means that deep, reliable insights into student-specific snacking behaviors, motivations, or barriers cannot be comprehensively extracted from this data.

### 5. Mentioned Brands & Products

The provided statistical output does not contain any specific brand or product mentions. Therefore, no insights can be extracted regarding specific brands or products or their associated sentiment.

### 6. Recommendations & Innovation Opportunities

Based on the observed data, the following opportunities for product development and messaging are suggested:

*   **Prioritize Convenience and Ready-to-Eat Formats:** Given the high preference for "Ready-to-Eat" (29.55%) and "Minimal Effort" (7.49%) snacks, there is a clear opportunity for products that offer maximum convenience for busy consumers and those snacking on-the-go.
*   **Address the "Healthy" Gap with Targeted Solutions:**
    *   **Focus on Protein/Energy and Sugar Reduction:** Develop products explicitly positioned as "High Protein / Energy" and "Sugar-Free / Low Sugar" to align with current consumer perceptions of health.
    *   **Overcome Barriers:** Innovate to offer healthy options that are:
        *   **Affordable:** Address "Price / Affordability" (9.01%) by developing cost-effective healthy snack lines.
        *   **Widely Available:** Improve distribution to tackle "Limited Availability" (3.7%).
        *   **Taste-Centric:** Prioritize taste to overcome "Taste Dissatisfaction" (3.32%), ensuring healthy options are also highly palatable.
*   **Cater to Both Functional and Emotional Needs:** Develop a diverse portfolio that addresses both "Hunger gap / tide over" (9.77%) and "Craving Fix" (7.58%). This could include functional snacks for sustained energy and indulgent options for emotional comfort.
*   **Explore Home and Family Snacking:** While "Solo" snacking is common, "At Home" (4.62%) and "With Children / Family" (2.88%) contexts suggest opportunities for larger pack sizes, family-friendly options, or snacks suitable for shared consumption.
*   **Target On-the-Go and School Contexts (with caution):** Despite limited data, the existence of "On the Go" (1.3%) and "At School" (0.66%) contexts indicates a need for portable, durable, and convenient snacks suitable for consumption outside the home, particularly for students. Further research would be beneficial to validate this segment's specific needs.